addons/mrp/models/mrp_production.py

```diff
@@ -2221,23 +2275,29 @@ def _pre_action_split_merge_hook(self, merge=False, split=False):
         products = set([(production.product_id, production.bom_id) for production in self])
         if len(products) > 1:
             raise UserError(_('You can only merge manufacturing orders of identical products with same BoM.'))
-        additional_raw_ids = self.mapped("move_raw_ids").filtered(lambda move: not move.bom_line_id)
-        additional_byproduct_ids = self.mapped('move_byproduct_ids').filtered(lambda move: not move.byproduct_id)
-        if additional_raw_ids or additional_byproduct_ids:
-            raise UserError(_("You can only merge manufacturing orders with no additional components or by-products."))
         if len(set(self.mapped('state'))) > 1:
             raise UserError(_("You can only merge manufacturing with the same state."))
-        if len(set(self.mapped('picking_type_id'))) > 1:
+        if len(self.picking_type_id) > 1:
             raise UserError(_('You can only merge manufacturing with the same operation type'))
+        additional_workorder_ids = self.workorder_ids.filtered(lambda workorder: not workorder.operation_id)
+        if additional_workorder_ids:
+            raise UserError(_("You can only merge manufacturing orders with no additional workorders."))
+        additional_raw_ids = self.move_raw_ids.filtered(lambda move: not move.bom_line_id)
+        additional_byproduct_ids = self.move_byproduct_ids.filtered(lambda move: not move.byproduct_id)
+        if self.bom_id and (additional_raw_ids or additional_byproduct_ids):
+            raise UserError(_("You can only merge manufacturing orders with no additional components or by-products."))
         # TODO explode and check no quantity has been edited
         return True
 
-    def _prepare_merge_orig_links(self):
+    def _prepare_merge_orig_links(self, merged_by='bom_line'):
         origs = defaultdict(dict)
         for move in self.move_raw_ids:
             if not move.move_orig_ids:
                 continue
-            origs[move.bom_line_id.id].setdefault('move_orig_ids', set()).update(move.move_orig_ids.ids)
+            if merged_by == 'bom_line':
```

Comment:
I don't see where `merged_by` could be different from `bom_line`

Reply:
Here I merge them by `product` since there is no bom
https://github.com/odoo/odoo/pull/117298/files#diff-6f4b1d411b3160a63b8260bf87166f0f20452b9b9fe3f04d28c3734302509f05R2138